### PR TITLE
Deploy to Docker Hub

### PR DESCRIPTION
**Pull request by rail**
The idea is to be able to deploy to CloudOps at some point. This adds an option to deploy images to Docker Hub, creating versioned (can be used for production deployment) and stable tags per channel (for staging and dev mostly).

I din't test the push step (inspected the vars in pdb), but what can possibly go wrong!? :)

### CODE REVIEW

File: lib/please_cli/please_cli/deploy.py [hunk 1]
```diff
@@ -463,3 +463,117 @@ def cmd_TASKCLUSTER_HOOK(ctx,
...
+    # remove the docker-image-mozilla- prefix and the extension
+    tag_base = project_basename.replace('docker-image-mozilla-', '').replace('.tar.gz', '')
+    # Puth the hash to the end of the tag
+    image_tag_versioned = '-'.join(reversed(tag_base.split('-', 1)))
+    for tag in (image_tag_versioned, image_tag):
+        click.echo(' => Uploading docker image `{}:{}` ... '.format(docker_repo, tag), nl=False)
+        with click_spinner.spinner():
+            push.registry.push(
+                spec,
+                please_cli.config.DOCKER_REGISTRY,
```
Comment (garbas) [inline lib/please_cli/please_cli/deploy.py]: I would make docker registry an option to this command which default to `please_cli.config.DOCKER_REGISTRY`

File: lib/please_cli/please_cli/deploy.py [hunk 2]
```diff
@@ -463,3 +463,117 @@ def cmd_TASKCLUSTER_HOOK(ctx,
...
+        project,
+        required=(
+            'DOCKER_USERNAME',
+            'DOCKER_PASSWORD',
+        ),
+        taskcluster_client_id=taskcluster_client_id,
+        taskcluster_access_token=taskcluster_access_token,
+    )
+
+    docker_username = secrets['DOCKER_USERNAME']
```
Comment (garbas) [inline lib/please_cli/please_cli/deploy.py]: I would make `docker_username` an command option that would default to `secrets['DOCKER_USERNAME']` and I would do the same for `docker_password`

File: lib/please_cli/please_cli/deploy.py [hunk 3]
```diff
@@ -463,3 +463,117 @@ def cmd_TASKCLUSTER_HOOK(ctx,
...
+    help='`nix` command',
+    )
+@click.option(
+    '--docker-repo',
+    required=True,
+    default=please_cli.config.DOCKER_REPO,
+    help='Docker repository.',
+    )
+@click.option(
+    '--image-tag-format',
```
Comment (garbas) [inline lib/please_cli/please_cli/deploy.py]: maybe lets add `--docker-` in front to make it obvious this is about docker image.
if turn this option to list of templates (`multiple=True`) and if you add `{hash}` and `{version}` as template parameters you could have a default that would do exactly what you hard-coded below.

**Comment by rail** [comment]
TBH, I tried to play around with unifying HEROKU, TASKCLATER_HOOK and DOCKERHUB, but the code just gets harder to read. Let's leave it as is until we have better refactoring ideas.

**Pull request merged**